### ISSUE TITLE
Question: reference function

### ISSUE THREAD

**Issue by peter1000** [MichaelHatherly/Lexicon.jl#43]
Is there an easy way to reference functions. Something similar to pythons documentation: **¶** Permanent link to this definition:  

e.g. https://docs.python.org/3/library/gzip.html#gzip.GzipFile.peek

**Comment by MichaelHatherly** [comment]
Not currently, but it would seem readthedocs adds an `id="..."` field to all the headers in generated output by just normalising the string to make it a valid url. That, combined with markdown links, is enough to link to definitions. It's a manual workaround unfortunately, but you can link directly to [an entry](http://docilejl.readthedocs.org/en/latest/api/Docile/index.html#documentoptions) it would seem.

A more automatic solution might be nice to have though.

**Comment by peter1000** [comment]
But that might be mkdocs theme dependent - would be nice to get something from within Docile/Lexicon/Markdown. I will ask over at mkdocs maybe they have some thoughts about it.

**Comment by MichaelHatherly** [comment]
Do we need specific syntax, such as http://sphinx-doc.org/markup/inline.html#ref-role, to denote which things to link? Probably. Unfortunately that's for only for rst, not markdown it would seem.

Interestingly rust seems to, from a quick glance, not try doing this automatically. See http://doc.rust-lang.org/1.0.0-beta/std/ and it's source https://github.com/rust-lang/rust/blob/6971727d45494b8362262bb41a45f2c0c5f162bc/src/libstd/lib.rs#L19. Might just be best to make the permalink accessible in the readthedocs/other outputs and just have the linking be done manually.

**Comment by peter1000** [comment]
http://doc.rust-lang.org/1.0.0-beta/std/  yes something like that. I was thinking that MkDocs makes section references: but I think only for H1, H2 maybe this could be extendet.

Otherwise: maybe one can do something automatic within Lexicon/Docile which would be nicer than writting all in the doc strings.

**Comment by MichaelHatherly** [comment]
Perhaps search each markdown AST for inline `Code` elements, for each one do a query of the current module docs for possible entries that match it and replace the `Code` element with a `Link` element pointing in the correct direction.

**Comment by MichaelHatherly** [comment]
Thought this sounded familiar... https://github.com/MichaelHatherly/Lexicon.jl/issues/2. Closing the other one since we've already got discussion here.

**Comment by peter1000** [comment]
d0ugal  from MkDocs was very kind to reply:
see: https://github.com/mkdocs/mkdocs/issues/421#issuecomment-89668828

On the otherhand having something within Lexicon/Docile is probably the prefered way - as you also output to html yourself.

**Comment by peter1000** [comment]
I will play a bit around with this issue and come back to you in a couple of days.

**Comment by peter1000** [comment]
question: just thought about: if we have cross link to functions ect.. we could do a module/function index similar to pythons: https://docs.python.org/3/py-modindex.html

What do you think about such and if -would that be a separate `function.md` or whatever name - or a section at the bottom of the `pkg.md`

**Comment by MichaelHatherly** [comment]
> we could do a module/function index similar to pythons

Sounds good, I like how it shows a short summary in the index per entry. 

> would that be a separate `function.md` or whatever name - or a section at the bottom of the `pkg.md`

Should be a configurable option I think. Sane default &ndash; a separate file in the generated folder that contains all generated api docs.

**Comment by peter1000** [comment]
multiple function with same name:

```julia

"""
testing1
"""
function writemime_test(io::IO, mime::MIME"text/md", md::Meta, mdstyle::Dict{Symbol, ByteString} = DEFAULT_MDSTYLE)
end

"""
testing2
"""
function writemime_test{category}(io::IO, mime::MIME"text/md", modname, obj, ent::Entry{category}, mdstyle::Dict{Symbol, ByteString} = DEFAULT_MDSTYLE)
end
```

MkDocs currently (if set with our new option style that to h1: (:objname        , "##"),) does the reference somewhat like this:

```
index.html#writemime_testioio-mimebasemultimediamimesymboltextmd-mdlexiconmetakeyword-mdstyledictsymbol-unionutf8string-asciistring

index.html#writemime_testcategoryioio-mimebasemultimediamimesymboltextmd-modname-obj-entdocileentrycategory
```

Personally I think it would be better to use indexes like:

```
index.html#writemime_test.1

index.html#writemime_test.2

```

Any thoughts on that?

**Comment by peter1000** [comment]
@MichaelHatherly I wanted to ask you already for the previous PR:

Do you have any fix ideas about julia code line length( official I think it is suggested 92 char).
I keep mine normally to about 160 char beause most editors have anyway line wrapping option.

For the other PR I did wrap acouple of lines as I was unsure about your policy.

**Comment by MichaelHatherly** [comment]
> Personally I think it would be better to use indexes like:

Yeah, those longer links start to get a bit messy. I'm fine with indices instead.

(Just a thought: It does mean that we've got to keep the indices the *same* between runs, which *might* not happen currently since Docile uses an `ObjectIdDict` for storage. Not sure whether it remains ordered.)

> Do you have any fix ideas about julia code line length

I try to keep lines less than 100 chars if possible.

**Comment by peter1000** [comment]
> I try to keep lines less than 100 chars if possible.

No problem I can wrap code.

> (Just a thought: It does mean that we've got to keep the indices the same between runs, which might not happen currently since Docile uses an ObjectIdDict for storage. Not sure whether it remains ordered.)

I did not look at docile till now: I start with just something simple and go from there.
Anyway, we only have to increment function if they are in the same `output.md` (which will be most of the time.

Current plan is somewhere around those lines

1. always add own anchor id
2. optional: add after or before or before the function a: ¶ with the link to the anchor
4. multiple functions on the same page: enumerate: anchor id
5. always generate a separate function index: add description full function sig.
6. Optional add short description: special Tag or First doc Line ?

**Comment by MichaelHatherly** [comment]
> ¶ with the link to the anchor (What do you think is it better after like rust and python?

After sounds fine to me. Going with what others do to avoid surprising new users is best.

> Optional add short description: special Tag or First doc Line ?

Lets go with just the first line. Makes people keep it short.

**Comment by peter1000** [comment]
> Lets go with just the first line. Makes people keep it short.
BTW: python doesn't do anything for function/methods index: https://docs.python.org/3/genindex-A.html

In our case the module index will not be of interest because most julia packages export only  1 or 2 modules.

**Comment by MichaelHatherly** [comment]
> In our case the module index will not be of interest because most julia packages export only 1 or 2 modules.

Since we can document `Module`s using Docile perhaps just having a module section in the index would be good rather than a whole separate index.

**Comment by peter1000** [comment]
just a question: because I do not know Dociles/Lexicon internals as well as you.

Is there an easy way to extract a method, function name from your parsed docs or do I have to use a string manipulation to extract only the name.

e.g.

```
save(file::AbstractString, modulename::Module)
```

What I mean: just to get the `save`

**Comment by MichaelHatherly** [comment]
`Docile.Interface.name` might be what you're looking for.

```
julia> Docile.Interface.name(save)
:save

julia> Docile.Interface.name(methods(save).defs) # A `Method` object.
:save

julia> Docile.Interface.name(Docile)
:Docile

julia> Docile.Interface.name(Docile.Entry)
:Entry
```

Do you want the `Function` `save` rather than the `Symbol` `:save`?

**Comment by MichaelHatherly** [comment]
`name` isn't actually exported, but Lexicon/Docile exchange quite a bit more than a single non-exported function so feel free to use it. It's unlikely to be going anywhere.

**Comment by peter1000** [comment]
Thanks - github updated the comment 5 ours later for me and in the meantime I had found it myself.
Am I correct that all objectnames can be retrived with `name` except the ones of macros using `macroname`

I got yesterday night a proto type working - does not look we need much changes.

I will still have to check out how to generate the: index page.

**Comment by peter1000** [comment]
The anchor I will add always so that the index page works: but the output of ** ¶** I think we should keep optional.

BTW: this will be "fix" and not like in pythons appear when one hovers over it.
That would be somthing one has to include in a MkDocs Theme: java script or so..

![untitled](https://cloud.githubusercontent.com/assets/6168860/7004835/94bc7b88-dc47-11e4-9a24-edb89f67ee2b.png)

**Comment by MichaelHatherly** [comment]
> Am I correct that all objectnames can be retrived with name except the ones of macros using macroname

Yes, macros are just a little awkward to deal with.

> this will be "fix" and not like in pythons appear when one hovers over it.

I'm fine with fixed permalinks for the moment.

The results looks good.

**Comment by peter1000** [comment]
just an unrelated code question 'cause I'm new to Julia:

https://github.com/MichaelHatherly/Lexicon.jl/blob/master/src/render/md.jl#L151

```julia
function footer(io::IO, ::MIME"text/md", doc::Metadata, mdstyle::Dict{Symbol, ByteString} = DEFAULT_MDSTYLE; mathjax = false)
    println(io, "")
end
```

Why do you need `::MIME"text/md", doc::Metadata, mdstyle::Dict{Symbol, ByteString} = DEFAULT_MDSTYLE; mathjax = false` if the function only does: `println(io, "")`

Well the `mdstyle` I added to be on the save side :)

**Comment by MichaelHatherly** [comment]
The structure of `md.jl` file was based on `html.jl` where a footer was needed I believe. Technically we could get rid of it, but then again we might want a footer at some point.

**Comment by peter1000** [comment]
EXAMPLE you might want to look at:

**MARKDOWN on Github**

For further notice look at the documentation:

@document(options...) https://github.com/peter1000/Lexicon_Test/blob/master/docs/api/Docile.md#@document.1

GENINDEX: https://github.com/peter1000/Lexicon_Test/blob/master/docs/api/genindex.md

**MkDocs page**
http://peter1000.github.io/Lexicon_Test/api/genindex/

**Comment by MichaelHatherly** [comment]
Yes, definitely heading in the right direction.

> Do you think we need to highlight each letter

Not sure, guess it would depend on the size of the package really. Perhaps for the moment just keeping it as a simple alphabetical list is fine.

**Comment by peter1000** [comment]
> a simple alphabetical list is fine.

sorting is fine but not sure what you think about macros 
e.g. simple sort
```
@query
EachEntry
Match
Query
QueryResults
calculate_score
doctest
filter
query
save
```

which means all macros will be together (maybe not bad).
or in the above case would it be better to have them like this.

e.g. special case macro
```
EachEntry
Match
Query
QueryResults
calculate_score
doctest
filter
@query
query
save
```

That would mean if there is only a macro it would also be listed below filter?

Anyway just asking your opinion.

**Comment by MichaelHatherly** [comment]
Perhaps each category should be sorted separately? And then order the categories something like:

* Modules
* Functions
* Methods (perhaps included with "Functions")
* Types
* Macros
* Constants

Or some other order... configurable?

**Comment by peter1000** [comment]
For now I settled with a simple Module index / and just sorted items for the index. Not sure how many packages will export internals and dor just a couple of exported functions too many submenu might be more than needed.

Anyway: here is a example: http://peter1000.github.io/Lexicon_Test/
(I used some subdirs to check if relative linking works)

**Extras:** I did implement it so that one could save multiple Pkg api into one common index.

e.g. Docile + Lexicon

```julia
using Lexicon, Docile
plmain = startgenidx("genindex.md"; idxheader = "#", idxmodname = "##")
save("api/Lexicon.md", Lexicon, plmain)
save("api/Docile.md", Docile, plmain)
savegenidx(plmain)
```
would produce:

```
├── api
│   ├── Docile.md
│   └── Lexicon.md
└── genindex.md     index to both
```

**Comment by peter1000** [comment]
I will do some code cleanup / write some docs and send a PR later tonight or tomorrow.

**Comment by MichaelHatherly** [comment]
Looks great, :+1: for the multi-Pkg bit!

**Comment by peter1000** [comment]
When you have time ould you look once again at: http://peter1000.github.io/Lexicon_Test/api/Lexicon/#lexicon

I rebuild it with MkDocs using another theme: which has a banner at the top.
Try clicking a `permalink` it will jump to high.

This is not a `actual` error and probably one can not fit all possible options (maybe there will be other themes with even bigger top banner but it still does not look too good.

For a test I tried to but the anchor higher above the hline but it is also not enough. Probably best to leaf it as it is and adjust the themes.

Anyway just wanted to let you know befor you consider to merge.

**Comment by peter1000** [comment]
Cause you are anyway bussy, I have an idea to help fix the above: probably I will send another  clean PR.

**Comment by peter1000** [comment]
Here the link to a fixed version: http://peter1000.github.io/Lexicon_Test_fixed/api/Lexicon/#lexicon  I will send you a clean new PR including this changes.

**Comment by peter1000** [comment]
hi, on the MkDocs google-group I got bac an interesting response.

===
 Waylan Limberg write:
So you want to have an ID so you can link to a header. But you want to provide your own custom value for the ID because the content of the header is is a function/code which doesn't result in a human readable auto-generated id. What you need is Markdown's Attribute List Extension:

https://pythonhosted.org/Markdown/extensions/attr_list.html

Then your Markdown would look like this:

    #### query(f::Function, sig::(Any..., ))   {#query.1}

And your HTML like this:

    <h4 id="query.1">query(f::Function, sig::(Any..., ))</h4>

Now the ID is not above the the header and your problem with the top banner should not exist.

Of course that doesn't give you the permalink. You could still create them manually, but the Table Of Contents Extension already offers that feature as an option you need to turn on. Unfortunately, mkdocs does not yet officially offer a way to set config options on Markdown extensions (coming soon see #435), but it is possible (see #366).

Until #435 is applied, you can do the following in your config file:

    markdown_extensions:
        - attr_list
        - toc(permalink=True)

After #435 is applied (version 1.0?), then your config file will need to look like this:

    markdown_extensions:
        attr_list: ~
        toc:
            permalink: True

In either case the Markdown input I provided above would then output the following HTML:

    <h4 id="query.1">query(f::Function, sig::(Any..., ))<a class="headerlink" href="#query.1" title="Permanent link">&para;</a></h4>

Hope that helps.

===

@MichaelHatherly what do you think - should I still look into `Markdown's Attribute List Extension`

My own opinion is that Lexicon should still do most of it independend from MkDocs - which will leave it open to render things with an other site generator or just have simple `markdown` generated index files.

Anyway maybe:  `Markdown's Attribute List Extension` would be of interest but I'm not sure if it works that you want to make it a requirement or if it would work with julia's Markdown at all.

**Comment by MichaelHatherly** [comment]
Thanks for forwarding this. I'd prefer not to become tied to any specific renderer; the output should remain independent of whatever someone chooses to use.

Julia's markdown is aiming to support http://spec.commonmark.org, perhaps there's mention of something similar to the `Attribute List Extension` there?

**Comment by peter1000** [comment]
closed https://github.com/MichaelHatherly/Lexicon.jl/pull/50 as there where too many changes in the mean time.

Will later start implementing the Permalink / Index.page - Probably in 2 smaller PR.

In the mean time `MkDocs` merge some suggestion to help working with the bootstrap themes. (https://github.com/mkdocs/mkdocs/commit/e19ae13b26a0a4b528a64cca5c4f35147786008b, https://github.com/mkdocs/mkdocs/issues/438)

**Comment by peter1000** [comment]
fix https://github.com/MichaelHatherly/Lexicon.jl/commit/b6583fbbf4afdd833b57ba0fb2cbe118f4908487
